`ClassCastException` when using `--import` with 0.11.3

With the recent release of 0.11.3 I noticed that a few of my plugins all started failing. The common thread of them seem to be that they are all using `--import`. For example you can replicate with the the `mill-github-dependency-graph`:

```
❯ mill --no-server --import ivy:io.chris-kipp::mill-github-dependency-graph::0.2.6 io.kipp.mill.github.dependency.graph.Graph/generate
[build.sc] [48/52] compile
[info] compiling 1 Scala source to /Users/ckipp/Documents/scala-workspace/bloop-config/out/mill-build/compile.dest/classes ...
[info] done compiling
[build.sc] [52/52] methodCodeHashSignatures
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at mill.main.client.IsolatedMillMainLoader.runMain(IsolatedMillMainLoader.java:58)
        at mill.main.client.MillClientMain.main(MillClientMain.java:64)
Caused by: java.lang.ClassCastException: class scala.collection.immutable.$colon$colon cannot be cast to class scala.Tuple2 (scala.collection.immutable.$colon$colon and scala.Tuple2 are in unnamed module of loader 'app')
        at scala.collection.Iterator$$anon$10.nextCur(Iterator.scala:594)
        at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:608)
        at scala.collection.immutable.List.prependedAll(List.scala:152)
        at scala.collection.immutable.List$.from(List.scala:684)
        at scala.collection.immutable.List$.from(List.scala:681)
        at scala.collection.IterableFactory$Delegate.from(Factory.scala:288)
        at scala.collection.immutable.Iterable$.from(Iterable.scala:35)
        at scala.collection.immutable.Iterable$.from(Iterable.scala:32)
        at scala.collection.IterableFactory$Delegate.from(Factory.scala:288)
        at scala.collection.IterableOps.flatMap(Iterable.scala:684)
        at scala.collection.IterableOps.flatMap$(Iterable.scala:684)
        at scala.collection.AbstractIterable.flatMap(Iterable.scala:933)
        at mill.resolve.Resolve.resolveNonEmptyAndHandle(Resolve.scala:230)
        at mill.resolve.Resolve.resolveNonEmptyAndHandle$(Resolve.scala:223)
        at mill.resolve.Resolve$Tasks$.resolveNonEmptyAndHandle(Resolve.scala:34)
        at mill.resolve.Resolve.$anonfun$resolve0$4(Resolve.scala:207)
        at scala.util.Either.map(Either.scala:382)
        at mill.resolve.Resolve.$anonfun$resolve0$3(Resolve.scala:206)
        at scala.collection.immutable.List.map(List.scala:246)
        at scala.collection.immutable.List.map(List.scala:79)
        at mill.resolve.Resolve.$anonfun$resolve0$2(Resolve.scala:205)
        at scala.collection.immutable.List.map(List.scala:246)
        at scala.collection.immutable.List.map(List.scala:79)
        at mill.resolve.Resolve.$anonfun$resolve0$1(Resolve.scala:204)
        at scala.util.Either.flatMap(Either.scala:352)
        at mill.resolve.Resolve.resolve0(Resolve.scala:203)
        at mill.resolve.Resolve.resolve0$(Resolve.scala:197)
        at mill.resolve.Resolve$Tasks$.resolve0(Resolve.scala:34)
        at mill.resolve.Resolve.resolve(Resolve.scala:194)
        at mill.resolve.Resolve.resolve$(Resolve.scala:189)
        at mill.resolve.Resolve$Tasks$.resolve(Resolve.scala:34)
        at mill.main.RunScript$.$anonfun$evaluateTasksNamed$1(RunScript.scala:24)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at mill.main.RunScript$.evaluateTasksNamed(RunScript.scala:24)
        at mill.runner.MillBuildBootstrap$.evaluateWithWatches(MillBuildBootstrap.scala:397)
        at mill.runner.MillBuildBootstrap.$anonfun$processFinalTargets$3(MillBuildBootstrap.scala:307)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at mill.runner.MillBuildBootstrap.processFinalTargets(MillBuildBootstrap.scala:307)
        at mill.runner.MillBuildBootstrap.evaluateRec(MillBuildBootstrap.scala:195)
        at mill.runner.MillBuildBootstrap.evaluate(MillBuildBootstrap.scala:47)
        at mill.runner.MillMain$.$anonfun$main0$6(MillMain.scala:213)
        at mill.runner.Watching$.watchLoop(Watching.scala:27)
        at mill.runner.MillMain$.$anonfun$main0$1(MillMain.scala:198)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at scala.Console$.withErr(Console.scala:193)
        at mill.api.SystemStreams$.$anonfun$withStreams$2(SystemStreams.scala:62)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at scala.Console$.withOut(Console.scala:164)
        at mill.api.SystemStreams$.$anonfun$withStreams$1(SystemStreams.scala:61)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at scala.Console$.withIn(Console.scala:227)
        at mill.api.SystemStreams$.withStreams(SystemStreams.scala:60)
        at mill.runner.MillMain$.main0(MillMain.scala:80)
        at mill.runner.MillMain$.main(MillMain.scala:58)
        at mill.runner.MillMain.main(MillMain.scala)
        ... 6 more
```

You can see it happening in CI [here](https://github.com/scalacenter/bloop-config/actions/runs/6220649898/job/16881074258). It is also happening with my ci-release plugins as you can see [here](https://github.com/scalacenter/bloop-config/actions/runs/6220649901/job/16881074182) with the same error.

Let me know if there is anything else I can provide.

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -25,7 +25,7 @@ case class Discover[T] private (
 ) {
   @deprecated
   def this(value: Map[Class[_], Seq[mainargs.MainData[_, _]]]) =
-    this(value.mapValues((Nil, _)).toMap)
+    this(value.view.mapValues((Nil, _)).toMap)
 }
 
 object Discover {

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -22,7 +22,10 @@ case class Discover[T] private (
       (Seq[String], Seq[mainargs.MainData[_, _]])
     ],
     dummy: Int = 0 /* avoid conflict with Discover.apply(value: Map) below*/
-)
+) {
+  def this(value: Map[Class[_], Seq[mainargs.MainData[_, _]]]) =
+    this(value.mapValues((Nil, _)).toMap)
+}
 
 object Discover {
   def apply2[T](value: Map[Class[_], (Seq[String], Seq[mainargs.MainData[_, _]])]): Discover[T] =

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -16,10 +16,13 @@ import scala.reflect.macros.blackbox
  * the `T.command` methods we find. This mapping from `Class[_]` to `MainData`
  * can then be used later to look up the `MainData` for any module.
  */
-case class Discover[T] private (val value: Map[
-  Class[_],
-  (Seq[String], Seq[mainargs.MainData[_, _]])
-], dummy: Int = 0 /* avoid conflict with Discover.apply(value: Map) below*/)
+case class Discover[T] private (
+    val value: Map[
+      Class[_],
+      (Seq[String], Seq[mainargs.MainData[_, _]])
+    ],
+    dummy: Int = 0 /* avoid conflict with Discover.apply(value: Map) below*/
+)
 
 object Discover {
   def apply2[T](value: Map[Class[_], (Seq[String], Seq[mainargs.MainData[_, _]])]): Discover[T] =

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -19,11 +19,14 @@ import scala.reflect.macros.blackbox
 case class Discover[T] private (val value: Map[
   Class[_],
   (Seq[String], Seq[mainargs.MainData[_, _]])
-])
+], dummy: Int = 0)
 
 object Discover {
   def apply2[T](value: Map[Class[_], (Seq[String], Seq[mainargs.MainData[_, _]])]): Discover[T] =
     new Discover[T](value)
+
+  def apply[T](value: Map[Class[_], Seq[mainargs.MainData[_, _]]]): Discover[T] =
+    new Discover[T](value.mapValues((Nil, _)).toMap)
 
   def apply[T]: Discover[T] = macro Router.applyImpl[T]
 

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -23,6 +23,7 @@ case class Discover[T] private (
     ],
     dummy: Int = 0 /* avoid conflict with Discover.apply(value: Map) below*/
 ) {
+  @deprecated
   def this(value: Map[Class[_], Seq[mainargs.MainData[_, _]]]) =
     this(value.mapValues((Nil, _)).toMap)
 }
@@ -31,6 +32,7 @@ object Discover {
   def apply2[T](value: Map[Class[_], (Seq[String], Seq[mainargs.MainData[_, _]])]): Discover[T] =
     new Discover[T](value)
 
+  @deprecated
   def apply[T](value: Map[Class[_], Seq[mainargs.MainData[_, _]]]): Discover[T] =
     new Discover[T](value.mapValues((Nil, _)).toMap)
 

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -34,7 +34,7 @@ object Discover {
 
   @deprecated
   def apply[T](value: Map[Class[_], Seq[mainargs.MainData[_, _]]]): Discover[T] =
-    new Discover[T](value.mapValues((Nil, _)).toMap)
+    new Discover[T](value.view.mapValues((Nil, _)).toMap)
 
   def apply[T]: Discover[T] = macro Router.applyImpl[T]
 

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -23,7 +23,7 @@ case class Discover[T] private (
     ],
     dummy: Int = 0 /* avoid conflict with Discover.apply(value: Map) below*/
 ) {
-  @deprecated
+  @deprecated("Binary compatibility shim", "Mill 0.11.4")
   def this(value: Map[Class[_], Seq[mainargs.MainData[_, _]]]) =
     this(value.view.mapValues((Nil, _)).toMap)
 }

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -19,7 +19,7 @@ import scala.reflect.macros.blackbox
 case class Discover[T] private (val value: Map[
   Class[_],
   (Seq[String], Seq[mainargs.MainData[_, _]])
-], dummy: Int = 0)
+], dummy: Int = 0 /* avoid conflict with Discover.apply(value: Map) below*/)
 
 object Discover {
   def apply2[T](value: Map[Class[_], (Seq[String], Seq[mainargs.MainData[_, _]])]): Discover[T] =

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -17,15 +17,27 @@ import scala.reflect.macros.blackbox
  * can then be used later to look up the `MainData` for any module.
  */
 case class Discover[T] private (
-    val value: Map[
+    value: Map[
       Class[_],
       (Seq[String], Seq[mainargs.MainData[_, _]])
     ],
     dummy: Int = 0 /* avoid conflict with Discover.apply(value: Map) below*/
 ) {
   @deprecated("Binary compatibility shim", "Mill 0.11.4")
-  def this(value: Map[Class[_], Seq[mainargs.MainData[_, _]]]) =
+  private[define] def this(value: Map[Class[_], Seq[mainargs.MainData[_, _]]]) =
     this(value.view.mapValues((Nil, _)).toMap)
+  // Explicit copy, as we also need to provide an override for bin-compat reasons
+  def copy[T](
+      value: Map[
+        Class[_],
+        (Seq[String], Seq[mainargs.MainData[_, _]])
+      ] = value,
+      dummy: Int = dummy /* avoid conflict with Discover.apply(value: Map) below*/
+  ): Discover[T] = new Discover[T](value, dummy)
+  @deprecated("Binary compatibility shim", "Mill 0.11.4")
+  private[define] def copy[T](value: Map[Class[_], Seq[mainargs.MainData[_, _]]]): Discover[T] = {
+    new Discover[T](value.view.mapValues((Nil, _)).toMap, dummy)
+  }
 }
 
 object Discover {

--- a/main/define/src/mill/define/Discover.scala
+++ b/main/define/src/mill/define/Discover.scala
@@ -32,7 +32,7 @@ object Discover {
   def apply2[T](value: Map[Class[_], (Seq[String], Seq[mainargs.MainData[_, _]])]): Discover[T] =
     new Discover[T](value)
 
-  @deprecated
+  @deprecated("Binary compatibility shim", "Mill 0.11.4")
   def apply[T](value: Map[Class[_], Seq[mainargs.MainData[_, _]]]): Discover[T] =
     new Discover[T](value.view.mapValues((Nil, _)).toMap)
 

--- a/main/resolve/src/mill/resolve/Resolve.scala
+++ b/main/resolve/src/mill/resolve/Resolve.scala
@@ -227,11 +227,11 @@ trait Resolve[T] {
       nullCommandDefaults: Boolean
   ): Either[String, Seq[T]] = {
     val rootResolved = ResolveCore.Resolved.Module(Segments(), rootModule.getClass)
-    val allPossibleNames = rootModule.millDiscover.value.values.flatMap(_._1).toSet
     val resolved =
       ResolveCore.resolve(rootModule, sel.value.toList, rootResolved, Segments()) match {
         case ResolveCore.Success(value) => Right(value)
         case ResolveCore.NotFound(segments, found, next, possibleNexts) =>
+          val allPossibleNames = rootModule.millDiscover.value.values.flatMap(_._1).toSet
           Left(ResolveNotFoundHandler(sel, segments, found, next, possibleNexts, allPossibleNames))
         case ResolveCore.Error(value) => Left(value)
       }

